pkg/executor/core/executor.go

```diff
@@ -62,87 +73,116 @@ func (executor Executor) VerifySubject(ctx context.Context, verifyParameters e.V
+	overallVerifySuccess := executor.PolicyEnforcer.OverallVerifyResult(ctx, verifierReports)
+	return types.VerifyResult{IsSuccess: overallVerifySuccess, VerifierReports: verifierReports}, nil
 }
 
-func (executor Executor) verifySubjectInternal(ctx context.Context, verifyParameters e.VerifyParameters) (types.VerifyResult, error) {
+// verifySubjectInternalWithoutDecision verifies the subject and returns result
+// without making decisions on the result.
+func (executor Executor) verifySubjectInternalWithoutDecision(ctx context.Context, verifyParameters e.VerifyParameters) ([]interface{}, error) {
 	subjectReference, err := utils.ParseSubjectReference(verifyParameters.Subject)
 	if err != nil {
-		return types.VerifyResult{}, err
+		return nil, err
 	}
 
 	desc, err := su.ResolveSubjectDescriptor(ctx, &executor.ReferrerStores, subjectReference)
 
 	if err != nil {
-		return types.VerifyResult{}, fmt.Errorf("resolving descriptor for the subject failed with error: %w", err)
+		return nil, fmt.Errorf("resolving descriptor for the subject failed with error: %w", err)
 	}
 
 	logrus.Infof("Resolve of the image completed successfully the digest is %s", desc.Digest)
 
 	subjectReference.Digest = desc.Digest
 
-	var verifierReports []interface{}
+	verifierReports := make([]interface{}, 0)
 	eg, errCtx := errgroup.WithContext(ctx)
 	var mu sync.Mutex
 
 	for _, referrerStore := range executor.ReferrerStores {
 		referrerStore := referrerStore
 		eg.Go(func() error {
 			var continuationToken string
-			wg := sync.WaitGroup{}
+			innerGroup, innerErrCtx := errgroup.WithContext(errCtx)
 			for {
 				referrersResult, err := referrerStore.ListReferrers(errCtx, subjectReference, verifyParameters.ReferenceTypes, continuationToken, desc)
 				if err != nil {
 					return err
 				}
 				continuationToken = referrersResult.NextToken
 				for _, reference := range referrersResult.Referrers {
-					wg.Add(1)
-					go func(reference ocispecs.ReferenceDescriptor) {
-						defer wg.Done()
-						if executor.PolicyEnforcer.VerifyNeeded(ctx, subjectReference, reference) {
-							verifyResult := executor.verifyReference(errCtx, subjectReference, reference, referrerStore)
+					if !executor.PolicyEnforcer.VerifyNeeded(innerErrCtx, subjectReference, reference) {
+						continue
+					}
+					reference := reference
+					innerGroup.Go(func() error {
+						if featureflag.UseRegoPolicy.Enabled {
+							verifyResult, err := executor.verifyReferenceForRegoPolicy(innerErrCtx, subjectReference, reference, referrerStore)
+							if err != nil {
+								logrus.Errorf("error while verifying reference %+v, err: %v", reference, err)
+								return err
+							}
+							mu.Lock() // locks the verifierReports List for write safety
+							defer mu.Unlock()
+							verifierReports = append(verifierReports, verifyResult)
+						} else {
+							verifyResult := executor.verifyReferenceForJSONPolicy(innerErrCtx, subjectReference, reference, referrerStore)
```

Comment:
general comment (non blocking): do we have an outlined plan for how we'd deprecate and then eventually get rid of the JSON policy? Will we every get rid of JSON policy?

Reply:
we would need to leave enough time for users to migrate from config policy to rego policy. @FeynmanZhou @toddysm how long do you think we should leave for migration?
For the migration, I plan to provide some equivalent rego policies for common config policies. And we probably need to have a performance evaluation on the rego policy approach as the new workflow will go through every report instead of failing early.